Source/Core/Core/AchievementManager.cpp

```diff
@@ -204,8 +206,85 @@ void AchievementManager::ActivateDeactivateRichPresence()
       nullptr, 0);
 }
 
+void AchievementManager::DoFrame()
+{
+  if (!m_is_game_loaded)
+    return;
+  m_threadguard = new Core::CPUThreadGuard(Core::System::GetInstance());
```

Comment:
Any reason to not just use `Core::RunAsCPUThread` directly? It would look much nicer than needing to shuffle around a RAII thread guard being used outside of a RAII context, like:

```cpp
Core::RunAsCPUThread([&] {
  rc_runtime_do_frame(
     &m_runtime,
     [](const rc_runtime_event_t* runtime_event) {
       AchievementManager::GetInstance()->AchievementEventHandler(runtime_event);
     },
     [](unsigned address, unsigned num_bytes, void* ud) {
       return AchievementManager::GetInstance()->MemoryPeeker(address, num_bytes, ud);
     },
     nullptr, nullptr);
});
```

same for the other cases where the guard is being heap allocated and then freed

Reply:
I'm looking into the syntax of this and I'm not entirely sure from it where I get the guard I need to pass into the memory calls.

Reply:
A lot of stuff that was only safe to run from the CPU Thread was recently changed to require a Core::CPUThreadGuard to be passed around as a reference to guarantee it's being run in a safe way.  Even though I think it's an eyesore, it was and remains pretty helpful for tracking down hidden issues.  In MenuBar.cpp, you can see an example of the intended way to use Core::CPUThreadGuard with the various symbol loading functions (GenerateSymbolsFromAddress, LoadSymbolMap, etc).  The idea is you want to make the guard a local variable of some higher level caller, then pass it around as a reference to the called functions which need it.  This avoid using any heap allocation.

Reply:
So instead of putting the CPUThreadGuard on the heap, storing a pointer to it in your class, and then needing to delete it to avoid memory leakage, just make the CPUThreadGuard a local variable of AchievementManager::DoFrame and pass it by reference to the AchievementManager::MemoryPeeker function that needs it.

Reply:
I also want to mention, the CPUThreadGuard has a GetSystem method which returns a reference to the Core::System singleton.  That should be your go-to way of accessing Core::System whenever you are using the CPUThreadGuard, rather than Core::System::GetInstance.

Reply:
Unfortunately I'm not sure I can modify MemoryPeeker's inputs - it needs to be a C function pointer and I can't capture anything on the lambda without breaking that.

Reply:
I apologize if I misunderstood the interface.  I have cloned your fork and will play with it to see if there is another solution.

Reply:
I do not know how to enable the achievements feature to test this, but here is my new suggestion.  Simply make the CPUThreadGuard a local variable constructed in AdvancementManager::MemoryPeeker.  Also, put Core::RunAsCPUThread where lioncache suggested, as this code can be run from the Video thread in dual core mode.  Perhaps also store a reference to Core::System in the AdvancementManager class (similar to how Memory::MemoryManager keeps a reference with m_system) to accessing Core::System later as a singleton.

Reply:
> I'm looking into the syntax of this and I'm not entirely sure from it where I get the guard I need to pass into the memory calls.

If it's only used for the MemoryPeeker function, you may just want to declare it as a normal stack variable inside the function instead of heap allocating it and deleting it after near every instance. Seems much simpler that way

Reply:
I'll admit testing is a little difficult right now for multiple reasons - I don't yet have the settings UI for enabling things properly, and we don't yet have testing data in the RetroAchievements server that would enable you to reach this point through normal operation. Still a work in progress.

Reply:
Now, I'm not trying to pull your leg, but my advice for accessing Core::System was moreso for a context where the CPUThreadGuard it is passed in as a parameter, which is what I was suggesting before I knew it was not applicable.  Most other code I've seen which must reach for the Core::System singleton use GetInstance() once and store it in a local reference named "system", and then further sub classes, accessed via stuff like GetMMU(), are stored grabbed once and stored as local references like "mmu".
Other than this coding preference nitpick, it looks much improved.

Reply:
Even in a context where the CPUThreadGuard is passed in by reference, you'll sometimes see the author also passed in System and other sub classes by reference alongside it.  So basically, these are just my thoughts.  Do what you think is best, and follow the DRY principle.